src/lsp4clj/server.clj

```diff
@@ -259,13 +199,91 @@
   protocols.endpoint/IEndpoint
   (start [this context]
     ;; Start receiving messages.
-    (let [pipeline (dispatch-input this context input-ch)]
-      ;; Wait to stop receiving messages.
+    (let [;; The language server sometimes needs to stop processing inbound
+          ;; requests and notifications.
+
+          ;; We do this automatically whenever we receive a notification, so
+          ;; that we are sure the language server processes didChange before
+          ;; moving on to other requests. But the language server can also
+          ;; decide to do it itself, even in a request, by processing everything
+          ;; synchronously instead of returning a future.
+
+          ;; In a request (or even in a notification) the language server can
+          ;; send a request to the client and block waiting for a response.
+
+          ;; It's possible -- even likely -- that the client will send other
+          ;; messages between the time that the server sent its request and the
+          ;; time that the client responds.
+
+          ;; All inbound messages -- requests, notifications, and responses --
+          ;; come in on a single stream.
+
+          ;; If the server blocks waiting for a response, we have to set aside
+          ;; the other inbound requests and notifications, so that we can get to
+          ;; the response. That is, while the server is blocking we cannot stop
+          ;; reading input. Otherwise, the server will end up in a deadlock,
+          ;; where it's waiting to receive a response off the input-ch but new
+          ;; messages aren't being put on the input-ch because the server is
+          ;; blocked. See
```

Comment:
I think this is not quite accurate. Messages keep being put on `input-ch`, but they are queued 

> where' it's waiting to receive a _response_ off the input-ch  **that is blocked by other messages before it in the queue**

Reply:
I think you might be confusing the `input-ch` and the new `client-initiated-in-ch`. The `input-ch` is a `(chan 1)`, fed by a `>!!` in `io-chan/input-stream->input-chan`, so it doesn't have room to queue anything. The new `client-initiated-in-ch` has a buffer and can queue messages. If you block between reading off the `input-ch`, as we were doing before, the feeder literally cannot put anything else onto it, not even into a queue.

That said, this comment is a bit hard to understand, as it's talking about a scenario that we used to have but no longer do. I'll edit it to talk more about the problem we're avoiding, and less about how specific configurations of queues can cause it.

Reply:
> I think you might be confusing the `input-ch` and the new `client-initiated-in-ch`. The `input-ch` is a `(chan 1)`, fed by a `>!!` in `io-chan/input-stream->input-chan`, so it doesn't have room to queue anything.

Users can use different implementations that `io-chan` (e.g. websockets), and could use different buffer strategies for the `input-ch`.

Anyway, thanks for improving this!